sdk/identity/azure-identity/azure/identity/_internal/shared_token_cache.py

```diff
@@ -0,0 +1,188 @@
+# ------------------------------------
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT License.
+# ------------------------------------
+import abc
+import os
+import sys
+
+
+from msal import TokenCache
+from azure.core.exceptions import ClientAuthenticationError
+from .._constants import KnownAuthorities
+
+try:
+    ABC = abc.ABC
+except AttributeError:  # Python 2.7, abc exists, but not ABC
+    ABC = abc.ABCMeta("ABC", (object,), {"__slots__": ()})  # type: ignore
+
+try:
+    from typing import TYPE_CHECKING
+except ImportError:
+    TYPE_CHECKING = False
+
+if TYPE_CHECKING:
+    # pylint:disable=unused-import,ungrouped-imports
+    from typing import Any, Iterable, List, Mapping, Optional
+    import msal_extensions
+    from azure.core.credentials import AccessToken
+    from .._internal import AadClientBase
+
+    CacheItem = Mapping[str, str]
+
+
+MULTIPLE_ACCOUNTS = """Multiple users were discovered in the shared token cache. If using DefaultAzureCredential, set
+the AZURE_USERNAME environment variable to the preferred username. Otherwise,
+specify it when constructing SharedTokenCacheCredential.\nDiscovered accounts: {}"""
+
+MULTIPLE_MATCHING_ACCOUNTS = """Found multiple accounts matching{}{}. If using DefaultAzureCredential, set environment
+variables AZURE_USERNAME and AZURE_TENANT_ID with the preferred username and tenant.
+Otherwise, specify them when constructing SharedTokenCacheCredential.\nDiscovered accounts: {}"""
+
+NO_ACCOUNTS = """The shared cache contains no signed-in accounts. To authenticate with SharedTokenCacheCredential, login
+through developer tooling supporting Azure single sign on"""
+
+NO_MATCHING_ACCOUNTS = """The cache contains no account matching the specified{}{}. To authenticate with
+SharedTokenCacheCredential, login through developer tooling supporting Azure single sign on.\nDiscovered accounts: {}"""
+
+NO_TOKEN = """Token acquisition failed for user '{}'. To fix, re-authenticate
+through developer tooling supporting Azure single sign on"""
+
+_PUBLIC_CLOUD_ALIASES = {"login.windows.net", "login.microsoft.com", "sts.windows.net"}
+
+
+def _account_to_string(account):
+    username = account.get("username")
+    home_account_id = account.get("home_account_id", "").split(".")
+    tenant_id = home_account_id[-1] if len(home_account_id) == 2 else ""
+    return "(username: {}, tenant: {})".format(username, tenant_id)
+
+
+def _filtered_accounts(accounts, username=None, tenant_id=None):
+    """yield accounts matching username and/or tenant_id"""
+
+    filtered_accounts = []
+    for account in accounts:
+        if username and account.get("username") != username:
+            continue
+        if tenant_id:
+            try:
+                _, tenant = account["home_account_id"].split(".")
+                if tenant_id != tenant:
+                    continue
+            except:  # pylint:disable=bare-except
```

Comment:
Why catch `BaseException`s here?

Reply:
Shorthand for `(AttributeError, KeyError, ValueError)`. Do you think it should be that specific, or just `Exception`?

Reply:
I would personally use `Exception` here, which will catch everything except keyboard interrupts, etc.

Reply:
Thanks, I was unaware of this subtlety around keyboard interrupts.